### PR TITLE
fix: struct type out of symbol error by duplicating it inside symbol table whenever required

### LINKED ISSUE #4192: bug: struct::m_derived_type cannot point outside of its symbol table
### Minimum Reproducible Example

```fortran
MODULE geom_module

   TYPE diag_type
      INTEGER :: len
   END TYPE diag_type

   TYPE(diag_type), DIMENSION(1) :: diag

END MODULE geom_module

PROGRAM test_program
    USE geom_module, ONLY: diag

   PRINT *, diag(1)

END PROGRAM test_program
```

### Output
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran -c ./examples/example.f90
ASR verify pass error: ASR verify: The symbol table was not found in the scope of `symtab`.
 --> ./examples/example.f90:3:4 - 6:1
  |
3 |       TYPE diag_type
  |       ^^^^^^^^^^^^^^...
...
  |
6 |    
  | ...^ failed here

ASR verify pass error: Struct::m_derived_type 'diag_type' cannot point outside of its symbol table, owner: geom_module
 --> ./examples/example.f90:7:4
  |
7 |    TYPE(diag_type), DIMENSION(1) :: diag
  |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ failed here


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

This bug was caught when trying to compile file `mkba_sweep.f90` of the SNAP package.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #4192

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3630,6 +3630,56 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         return 1; // default
     }
 
+    void create_and_replace_structType() {
```
Comment (gxyd) [inline src/lfortran/semantics/ast_common_visitor.h:3633]: I'm curious, if the existing `duplicate_Struct` can work for us: https://github.com/lfortran/lfortran/blob/829526d765a91ac6a508c99b35bd95ed3c4b6df4/src/libasr/asr_utils.h#L4315?

**Comment by gxyd** [comment]
Not sure how important is to have the below program working correctly, but it fails with this change:
```f90
MODULE geom_module
    TYPE diag_type
       INTEGER :: len = 10
    END TYPE diag_type

    TYPE(diag_type), DIMENSION(1) :: diag
END MODULE geom_module
 
PROGRAM test_program
    USE geom_module, only: diag
    PRINT *, diag(1)         ! this prints `0` with your change, expected 10 (as does GFortran)
END PROGRAM test_program
```

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:3633]: No, the current function extends `BaseWalkVisitor` with a special handling for `ArrayItem` and `StructType` and thus both are different. Also, we shall remove this `duplicate_Struct` function from `ASRUtils` and rather use `ASRUtils::SymbolDuplicator sd(al);`

**Comment by Pranavchiku** [comment]
That is I believe trouble in `ASRUtils::symbolduplicator` which is not catering for `len = 10`. That is fine, easy to fix. Also, if not required for SNAP, I'll prefer to do it later.

**Comment by certik** [comment]
@Pranavchiku awesome, thank you so much for fixing it!

Can you explain to me what the new function is doing? I can see a visitor there: what exactly is it visiting? I can see it is visiting the current scope, so if the struct is in a module, is it visiting the whole module? So every time we declare a struct, it will visit the whole module?

I am a little worried about performance of compilation.

**Review by certik — APPROVED** [review]
I think the change itself is fine.

Can there be a big compilation performance penalty here? That's my only worry. It would be good to check.

**Comment by certik** [comment]
You can use `./build_release.sh` to build in Release mode, and let's compile SNAP and measure compilation time.

**Comment by certik** [comment]
I am getting 0.209s with this PR:
```
time make -j8 FORTRAN=lfortran FFLAGS= MPI=no OPENMP=no snap_main.o
```
and this diff and lf7 branch:
```diff
--- a/src/snap_main.f90
+++ b/src/snap_main.f90
@@ -1,4 +1,8 @@
-PROGRAM snap_main
+module snap_main
+
+contains
+
+subroutine main()

 !-----------------------------------------------------------------------
 !
@@ -251,4 +255,6 @@ PROGRAM snap_main
 !_______________________________________________________________________
 !_______________________________________________________________________

-END PROGRAM snap_main
+end subroutine
+
+END module
```

**Comment by certik** [comment]
I tried to revert the workaround for struct in SNAP, however I am still getting a segfault:
```
make: *** [Makefile:180: octsweep.o] Segmentation fault
```

**Comment by Pranavchiku** [comment]
> Can you explain to me what the new function is doing?

This function visits the built `program`, `subroutine` or `function` and checks if we have an `ArrayItem` with `StructType` as type. If yes, then it duplicates `Struct` symbol into `current_scope` and uses that instead.

> Can there be a big compilation performance penalty here?

This will visit each statement of program again.

**Comment by Pranavchiku** [comment]
> Here is the error message I am getting in Debug mode:

I expect this to happen. We need to create MRE and iteratively improve it. I believe there will be a few cases and that will be fine.

**Comment by Pranavchiku** [comment]
If possible, can you please post `cell_id_type` struct? I believe we need to add `duplicate_type` at correct places to assure the node `StructType` is different from what we have in module.

**Comment by Pranavchiku** [comment]
MRE:

```fortran
module geomm_module


  TYPE cell_id_type
    INTEGER:: ic, j, k
  END TYPE cell_id_type

  TYPE diag_type
    INTEGER:: len
    TYPE(cell_id_type), ALLOCATABLE, DIMENSION(:) :: cell_id
  END TYPE diag_type

  TYPE(diag_type), ALLOCATABLE, DIMENSION(:) :: diag

end module

program main
        use geomm_module
        print *, diag(1)%cell_id(1)%ic
end program
```

**Comment by Pranavchiku** [comment]
Another MRE:

```fortran
module geomm_module


  TYPE cell_id_type
    INTEGER:: ic, j, k
  END TYPE cell_id_type

  TYPE diag_type
    INTEGER:: len
    TYPE(cell_id_type), ALLOCATABLE, DIMENSION(:) :: cell_id
  END TYPE diag_type

  TYPE(diag_type), ALLOCATABLE, DIMENSION(:) :: diag

end module

MODULE mkba_sweep_module

  USE geomm_module, ONLY: diag
  
  CONTAINS

  SUBROUTINE mkba_sweep ()
    print *, diag(1)%cell_id(1)%ic
  END SUBROUTINE mkba_sweep


END MODULE mkba_sweep_module
```

This is fixed.

**Comment by Pranavchiku** [comment]
Now SNAP works for me.

**Comment by Pranavchiku** [comment]
While debugging the integration test failure, I realised that we do not need the verify check as `StructType` is declared globally in generated LLVM. Thus https://github.com/lfortran/lfortran/pull/4668.

**Comment by Pranavchiku** [comment]
For this PR there is only 1 integration test failing, I am trying to fix it.

**Comment by Pranavchiku** [comment]
With this PR we can revert https://github.com/certik/SNAP/commit/d82b86772758258c7e783ef7402c0ef1240171d9 commit from SNAP. It works fine.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -3630,6 +3630,123 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+
+                void visit_StructType( const ASR::StructType_t& x ) {
+                    if ( sem >= 0 ) {
+                        ASR::StructType_t& xx = const_cast<ASR::StructType_t&>(x);
+                        if ( ASRUtils::symbol_parent_symtab(xx.m_derived_type)->counter != local_current_scope->counter) {
+                            if ((local_current_scope->resolve_symbol(ASRUtils::symbol_name(xx.m_derived_type)) == nullptr)) {
+                                ASRUtils::SymbolDuplicator sd(al);
+                                sd.duplicate_symbol(xx.m_derived_type, local_current_scope);
+                            }
+                            xx.m_derived_type = local_current_scope->resolve_symbol(ASRUtils::symbol_name(xx.m_derived_type));
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:3727]: For ExternalSymbol, one shall create it here.

**Comment by certik** [comment]
I verified that this PR fixes SNAP. I think this fix is close to the best we can do right now. Later we'll refactor to not link to a symbol table at all.

**Pull request merged**